### PR TITLE
util: decode bytes inplace

### PR DESCRIPTION
**Pull request by Connor1996**
## What have you changed? (mandatory)
- add `decode_bytes_inplace()` to save mallocation
- add `take_raw()` for `Key` which use `decode_bytes_inplace()`
- call `take_raw()` instead of `raw()` in some code place

## What are the type of the changes? (mandatory)
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
unit-test

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)
No

## Does this PR affect tidb-ansible update? (mandatory)
No

## Benchmark result if necessary (optional)
I write a micro bench for two decode_bytes_inplace implementation.
one for copying bytes one by one
```rust
 for i in write_offset..(write_offset + ENC_GROUP_SIZE){
     data[i] = data[read_offset];
     read_offset += 1;
}
```
the other for memmoving 8 bytes in one shot
```rust
ptr::copy(
    data.as_ptr().offset(read_offset as isize),
    data.as_mut_ptr().offset(write_offset as isize),
    ENC_GROUP_SIZE - pad_size
);
```
here is the result
```
test util::codec::bytes::tests::bench_decode                 ... bench:   3,291,861 ns/iter (+/- 234,919)
test util::codec::bytes::tests::bench_decode_inplace_copy    ... bench:   4,772,352 ns/iter (+/- 1,063,849)
test util::codec::bytes::tests::bench_decode_inplace_memmove ... bench:   2,591,718 ns/iter (+/- 557,794)
```

### CODE REVIEW

File: src/raftstore/store/region_snapshot.rs [hunk 1]
```diff
@@ -532,32 +532,32 @@ mod tests {
         let mut iter = Cursor::new(snap.iter(IterOption::default()), ScanMode::Mixed);
         assert!(
             !iter
-                .reverse_seek(&Key::from_encoded(b"a2".to_vec()), &mut statistics)
+                .reverse_seek(&Key::from_encoded_slice(b"a2"), &mut statistics)
```
Comment (breeswish) [inline src/raftstore/store/region_snapshot.rs]: Off topic.. is Key::from_encoded correct in this test case? It seems to be "from_raw" 🤣

**Comment by breeswish** [comment]
What's the difference between decode_bytes_in_place_memmove and decode_bytes_in_place? BTW I think it should accept a &mut so that it is "in place".

File: src/raftstore/store/region_snapshot.rs (hunk 1 above)
Comment (Connor1996) [reply src/raftstore/store/region_snapshot.rs]: oh, I will clean these. if it should be `from_raw()` I will fix it in other PR.

**Comment by Connor1996** [comment]
@breeswish These are two implementation and you can see the bench result for the difference. I'm working on cleaning this and add some comment.

File: src/util/codec/bytes.rs [hunk 2]
```diff
@@ -199,11 +210,83 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
...
+        } else {
+            (ENC_MARKER - marker) as usize
+        };
+
+        // no padding, just move 8 bytes to write_offset
+        if pad_size == 0 {
+            unsafe {
+                // it is semantically equivalent to C's memmove()
+                // and the src and dest may overlap
+                ptr::copy(
```
Comment (zhangjinpeng1987) [inline src/util/codec/bytes.rs]: If read_offset == write_offset, don't need to copy.

**Comment by Connor1996** [comment]
> BTW I think it should accept a &mut so that it is "in place".
if this function return Err, &mut may let the vec incomplete decode and caller may misuse the vec

File: src/util/codec/bytes.rs [hunk 3]
```diff
@@ -199,11 +210,84 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
...
         }
+        // data will point to following unencoded bytes, maybe timestamp
         *data = &data[offset..];
         return Ok(key);
     }
 }
 
+/// `decode_bytes_in_place` decodes bytes which is encoded by `encoded_bytes` before just in place without malloc.
+/// Please use this instead of `decode_bytes` if possible.
+pub fn decode_bytes_in_place(mut data: Vec<u8>, desc: bool) -> Result<Vec<u8>> {
```
Comment (zhangjinpeng1987) [inline src/util/codec/bytes.rs]: I prefer use &mut data as the argument.

File: src/util/codec/bytes.rs (hunk 2 above)
Comment (MyonKeminta) [reply src/util/codec/bytes.rs]: I've checked glibc and It seems it didn't skip the condition where `dst == src`. I agree that you do not copy when read_offset == write_offset.

File: src/util/codec/bytes.rs [hunk 4]
```diff
@@ -162,34 +163,44 @@ pub fn decode_compact_bytes(data: &mut BytesSlice) -> Result<Vec<u8>> {
     Err(Error::unexpected_eof())
 }
 
+
+/// `decode_bytes` decodes bytes which is encoded by `encoded_bytes` before.
```
Comment (breeswish) [inline src/util/codec/bytes.rs]: typo: `encode_bytes`.

**Comment by breeswish** [comment]
Interesting.. ptr::copy is faster than directly copy as u64:

```rust
        unsafe {
            let v = *(&data[read_offset] as *const u8 as *const u64);
            *(&mut data[write_offset] as *mut u8 as *mut u64) = v;
        }
```

@zhangjinpeng1987

**Comment by breeswish** [comment]
I found out the reason of slow is because of the boundary check in [] operator.

**Comment by breeswish** [comment]
I successfully made some even more improvements over your code, 30% faster:

```rust
pub fn decode_bytes_in_place4(data: &mut Vec<u8>) -> Result<(), &'static str> {
    unsafe {
        let mut write_offset = 0;
        let mut read_offset = 0;
        let len = data.len();
        if len % 9 != 0 || len == 0 {
            return Err("Unexpected EOF");
        }
        if *data.get_unchecked(len - 1) == 255 {
            // Last byte must be a > 0 padding size.
            return Err("Padding error");
        }
        loop {
            let marker_position = read_offset + ENC_GROUP_SIZE;
            let next_position = marker_position + 1;

            let chunk: u64 = *(data.as_ptr().offset(read_offset as isize) as *const u64);
            *(data.as_mut_ptr().offset(write_offset as isize) as *mut u64) = chunk;

            read_offset = next_position;
            write_offset += ENC_GROUP_SIZE;
            let pad_size = (ENC_MARKER - data.get_unchecked(marker_position)) as usize;
            if pad_size > 0 {
                if pad_size > ENC_GROUP_SIZE {
                    return Err("Padding error");
                }
                if data[marker_position - pad_size..marker_position] != ENC_PADDING[..pad_size] {
                    return Err("Padding error");
                }
                data.set_len(write_offset - pad_size);
                return Ok(());
            }
        }
    }
}
```

(notice that I was using `&'static str` because I was writing a standalone test outside TiKV. You need to change it back to our `Error` type)

Result:

With valid padding check (30% faster than this PR):

```
test tests::bench_decode          ... bench:      48,991 ns/iter (+/- 19,410)
test tests::bench_decode_inplace  ... bench:      15,327 ns/iter (+/- 2,364)
test tests::bench_decode_inplace4 ... bench:      10,678 ns/iter (+/- 2,000)
```

Without valid padding check (42% faster than this PR):

```
test tests::bench_decode          ... bench:      42,859 ns/iter (+/- 14,405)
test tests::bench_decode_inplace  ... bench:      15,235 ns/iter (+/- 2,181)
test tests::bench_decode_inplace4 ... bench:       8,836 ns/iter (+/- 1,554)
```

The core idea is to eliminate `if` statements, for example, removing "desc" helps a lot.

**Comment by breeswish** [comment]
Use `ptr::write` (write aligned memory) instead of `*xxxx = chunk` can be even more faster (benchmark without valid padding check):

```
test tests::bench_decode_inplace4 ... bench:       8,296 ns/iter (+/- 1,978)
```

**Comment by zhangjinpeng1987** [comment]
@breeswish There is something wrong with your code, because it is decode in place, so the source and destination can overlap, so should use memmove.

**Comment by breeswish** [comment]
It reads memory to a variable first, then writes to another memory. I think the overlapping does not matter. Otherwise unit tests will fail.

**Comment by Connor1996** [comment]
@breeswish 
when I only replace `ptr::copy()` with this, seems that it doesn't make any progress.
```
let chunk: u64 = *(data.as_ptr().offset(read_offset as isize) as *const u64);
            *(data.as_mut_ptr().offset(write_offset as isize) as *mut u64) = chunk;
```

**Comment by Connor1996** [comment]
@breeswish And how is your bench code written

**Comment by breeswish** [comment]
@Connor1996 As said previously, most improvements are caused by eliminating `if` statements, for example, remove the `desc` option and remove most of the branches in loops.

**Comment by breeswish** [comment]
Also my benchmark is like:

```rust
    #[bench]
    fn bench_decode_inplace4(b: &mut Bencher) {
        let key = [b'x'; 100000];
        let encoded = encode_bytes(&key);
        b.iter(|| {
            let mut value = test::black_box(encoded.clone());
            test::black_box(decode_bytes_in_place4(&mut value));
        });
    }
```

By moving encode_bytes to outside and using a clone instead, the result will be more relevant.

**Comment by Connor1996** [comment]
new bench result
```
test util::codec::bytes::tests::bench_decode                                   ... bench:       4,486 ns/iter (+/- 4,504)
test util::codec::bytes::tests::bench_decode_inplace                      ... bench:       1,383 ns/iter (+/- 117)
test util::codec::bytes::tests::bench_decode_inplace_write            ... bench:       1,406 ns/iter (+/- 556)
test util::codec::bytes::tests::bench_decode_inplace_write_small  ... bench:          38 ns/iter (+/- 4)
test util::codec::bytes::tests::bench_decode_inplace_small            ... bench:          36 ns/iter (+/- 2)
test util::codec::bytes::tests::bench_decode_small                          ... bench:          69 ns/iter (+/- 3)
```

@breeswish seems that `ptr::write` doesn't make progress either, the prev result maybe cause by Deviation. And I didn't accept the checking code cause it didn't cover the case that there are some timestamp bytes after encoded bytes.

**Comment by breeswish** [comment]
Wow. How would timestamped bytes affect the correctness for current functions? I previously thought that timestamps are always truncated before passing to the decode function. I tested my function over current codec test cases and it passed. If this is a scenario we need to consider, please add tests to ensure that.

**Comment by Connor1996** [comment]
current function decodes from front to last padding bytes and just ignores rest bytes. Actually we call `raw()/take_raw()` after `truncate()`, but it maybe misused, I will add test for it.

File: src/util/codec/bytes.rs [hunk 5]
```diff
@@ -199,11 +214,74 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
...
+        unsafe {
+            // the last byte in decode unit is for marker which indicates pad size
+            let marker = *data.get_unchecked(marker_offset);
+            let pad_size = if desc {
+                marker as usize
+            } else {
+                (ENC_MARKER - marker) as usize
+            };
+
+            if pad_size > 0 {
```
Comment (zhangjinpeng1987) [inline src/util/codec/bytes.rs]: How about `pad_size==0` for the last group?

**Comment by breeswish** [comment]
Oh, got it. @Connor1996

File: src/util/codec/bytes.rs [hunk 6]
```diff
@@ -199,11 +214,74 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
...
+                data.as_mut_ptr().offset(write_offset as isize),
+                ENC_GROUP_SIZE,
+            );
+        }
+        write_offset += ENC_GROUP_SIZE;
+        read_offset = next_offset;
+
+        unsafe {
+            // the last byte in decode unit is for marker which indicates pad size
+            let marker = *data.get_unchecked(marker_offset);
```
Comment (zhangjinpeng1987) [inline src/util/codec/bytes.rs]: Please add comment why not use data[marker_offset] but use get_unchecked.

File: src/util/codec/bytes.rs [hunk 7]
```diff
@@ -199,11 +214,75 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
...
+                data.as_mut_ptr().offset(write_offset as isize),
+                ENC_GROUP_SIZE,
+            );
+        }
+        write_offset += ENC_GROUP_SIZE;
+        read_offset = next_offset;
+
+        unsafe {
+            // the last byte in decode unit is for marker which indicates pad size
+            // for the sake of performancce, use get_unchecked() to save bound check
```
Comment (BusyJay) [inline src/util/codec/bytes.rs]: I don't think bound check exists anymore as L232 has proven it to the compiler.

**Comment by breeswish** [comment]
> current function decodes from front to last padding bytes and just ignores rest bytes. Actually we call raw()/take_raw() after truncate(), but it maybe misused, I will add test for it.

Please open an issue to track this misuse. As you can see, this kind of misuse prevents us from using a more efficient implementation 🤣 and it is likely to cause mistakes in future because we know nothing about this when looking at the codec.

File: src/util/codec/bytes.rs (hunk 7 above)
Comment (Connor1996) [reply src/util/codec/bytes.rs]: maybe it can not make sure `data.len()` whether change or not, and from bench result showed, uncheck does make sense.

Comment (BusyJay) [reply src/util/codec/bytes.rs]: Seems `+ 1` confuse compiler. Using `marker_offset >= data.len()` can omit the bound check.

File: src/util/codec/bytes.rs [hunk 8]
```diff
@@ -199,11 +214,75 @@ pub fn decode_bytes(data: &mut BytesSlice, desc: bool) -> Result<Vec<u8>> {
...
+                    return Err(Error::KeyPadding);
+                }
+
+                // check the padding pattern whether validate or not
+                let padding_slice = if desc {
+                    &ENC_DESC_PADDING[..pad_size]
+                } else {
+                    &ENC_ASC_PADDING[..pad_size]
+                };
+                if &data[write_offset - pad_size..write_offset] != padding_slice {
```
Comment (BusyJay) [inline src/util/codec/bytes.rs]: Have you tried to use `libc::memcmp`?

**Comment by breeswish** [comment]
Circle CI failed, please `make dev`.

File: src/util/codec/bytes.rs (hunk 8 above)
Comment (Connor1996) [reply src/util/codec/bytes.rs]: &[u8] it does use memcmp to cmp
```
// Use memcmp for bytewise equality when the types allow
impl<A> SlicePartialEq<A> for [A]
    where A: PartialEq<A> + BytewiseEquality
{
    fn equal(&self, other: &[A]) -> bool {
        if self.len() != other.len() {
            return false;
        }
        if self.as_ptr() == other.as_ptr() {
            return true;
        }
        unsafe {
            let size = mem::size_of_val(self);
            memcmp(self.as_ptr() as *const u8,
                   other.as_ptr() as *const u8, size) == 0
        }
    }
}
```

Comment (Connor1996) [reply src/util/codec/bytes.rs]: &[u8] it does use memcmp to cmp
```
// Use memcmp for bytewise equality when the types allow
impl<A> SlicePartialEq<A> for [A]
    where A: PartialEq<A> + BytewiseEquality
{
    fn equal(&self, other: &[A]) -> bool {
        if self.len() != other.len() {
            return false;
        }
        if self.as_ptr() == other.as_ptr() {
            return true;
        }
        unsafe {
            let size = mem::size_of_val(self);
            memcmp(self.as_ptr() as *const u8,
                   other.as_ptr() as *const u8, size) == 0
        }
    }
}
```

File: src/util/codec/bytes.rs (hunk 7 above)
Comment (Connor1996) [reply src/util/codec/bytes.rs]: you are right, `marker_offset >= data.len()` makes sense.

**Comment by breeswish** [comment]
@Connor1996 In test cases there is no assurance for this kind of usage at all. Even after you adds these new test cases I bet my implementation will pass it 🤣 (and fail the code in elsewhere)

**Comment by Connor1996** [comment]
I have already add the tests, and your implementation will regard it as Unexpected EOF or PaddingError

**Pull request merged**